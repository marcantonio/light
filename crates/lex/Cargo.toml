[package]
name = "lex"
version = "0.0.0"
edition = "2021"

[dependencies]
itertools = "0.10.5"
serde = { version = "1.0.136", features = ["derive"] }

common = { path = "../common" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }
