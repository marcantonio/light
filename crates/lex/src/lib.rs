use serde::Serialize;
use std::iter::Peekable;

use common::Operator;
pub use token::{Token, TokenType};

#[cfg(test)]
mod tests;
pub mod token;

pub type LexResult = std::result::Result<Token, LexError>;

pub struct Lex {
    stream: Peekable<StreamIter<char>>,
    pub tokens: Vec<Token>,
}

impl Lex {
    pub fn new(input: &str) -> Self {
        Lex { stream: StreamIter::new(input).peekable(), tokens: vec![] }
    }

    // Scan all input
    pub fn scan(mut self) -> Result<Vec<Token>, LexError> {
        loop {
            let token = self.lex()?;
            if token.is_eof() {
                break;
            }
            self.tokens.push(token);
        }
        Ok(self.tokens)
    }

    // Recursively process enough characters to produce one token
    fn lex(&mut self) -> LexResult {
        use TokenType::*;

        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => unreachable!("can't lex nothing"),
        };

        // Inject a semicolon if certain tokens occur at the end of the line or
        // EOF. If EOF, make sure the context is right.
        if cur.value == '\n' && self.should_add_semicolon() {
            return Ok(Token::new(Semicolon(true), cur.line, cur.column));
        } else if cur.is_eof() {
            if self.should_add_semicolon() {
                let semi = match self.tokens.last() {
                    Some(t) => Token::new(Semicolon(true), t.line, t.column + 1),
                    None => Token::default(),
                };
                self.tokens.push(semi);
            }
            return Ok(Token::new(Eof, cur.line, cur.column));
        }

        // Ignore whitespace
        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Single line comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Keywords, types, and identifiers
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "fn" => Fn,
                "let" => Let,
                "for" => For,
                "if" => If,
                "else" => Else,
                "extern" => Extern,
                "true" => Bool(true),
                "false" => Bool(false),
                "struct" => Struct,
                "module" => Module,
                _ => Ident(identifier),
            };

            return Ok(Token::new(tt, cur.line, cur.column));
        }

        // Literal numbers
        if cur.value.is_ascii_digit() {
            let mut n = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    n.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return Ok(Token::new(Num(n), cur.line, cur.column));
        }

        // Literal char
        if cur == '\'' {
            let mut ch = String::new();
            let next = self.stream.next().unwrap_or_else(|| unreachable!("lexed None when looking for char"));

            match next.value {
                // Control characters
                '\\' => {
                    if let Some(next) = self.stream.next() {
                        match next.value {
                            'n' => ch = String::from("\n"),
                            't' => ch = String::from("\t"),
                            '\'' => ch = String::from("'"),
                            c => {
                                return Err(LexError::from((
                                    format!("Invalid character control sequence: `\\{}`", c),
                                    next,
                                )))
                            },
                        }
                    }
                },
                // EOF
                '\0' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`, got `EOF`".to_string(),
                        cur,
                    )));
                },
                '\'' => return Err(LexError::from(("Character literal can't be empty".to_string(), cur))),

                // Everything else
                c => ch = String::from(c),
            }

            // Check for closing '\''
            let last =
                self.stream.next().unwrap_or_else(|| unreachable!("fatal: lexed None when looking for `'`"));
            match last.value {
                '\'' => (),
                '\0' | '\n' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`".to_string(),
                        last,
                    )));
                },
                _ => {
                    return Err(LexError::from((
                        format!("Invalid character sequence: `'{}{}'`", ch, last.value),
                        last,
                    )));
                },
            }

            return Ok(Token::new(Char(ch), cur.line, cur.column));
        }

        // Multi-character operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Eq), cur.line, cur.column));
                },
                '!' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::NotEq), cur.line, cur.column));
                },
                '>' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::GtEq), cur.line, cur.column));
                },
                '<' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::LtEq), cur.line, cur.column));
                },
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::And), cur.line, cur.column));
                },
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Or), cur.line, cur.column));
                },
                '+' if next == &'+' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Inc), cur.line, cur.column));
                },
                '+' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::AddEq), cur.line, cur.column));
                },
                '-' if next == &'-' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Dec), cur.line, cur.column));
                },
                '-' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::SubEq), cur.line, cur.column));
                },
                '-' if next == &'>' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::RetType), cur.line, cur.column));
                },
                '*' if next == &'*' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::Pow), cur.line, cur.column));
                },
                '*' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::MulEq), cur.line, cur.column));
                },
                '/' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::new(Op(Operator::DivEq), cur.line, cur.column));
                },
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => Op(Operator::Add),
            '-' => Op(Operator::Sub),
            '*' => Op(Operator::Mul),
            '/' => Op(Operator::Div),
            '>' => Op(Operator::Gt),
            '<' => Op(Operator::Lt),
            '!' => Op(Operator::Not),
            '=' => Op(Operator::Assign),
            '&' => Op(Operator::BitAnd),
            '^' => Op(Operator::BitXor),
            '|' => Op(Operator::BitOr),
            '.' => Dot,
            '}' => CloseBrace,
            ']' => CloseBracket,
            ')' => CloseParen,
            ':' => Colon,
            ',' => Comma,
            '{' => OpenBrace,
            '[' => OpenBracket,
            '(' => OpenParen,
            ';' => Semicolon(false),
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            },
        };

        Ok(Token::new(tt, cur.line, cur.column))
    }

    // Add a semicolon for these tokens
    fn should_add_semicolon(&self) -> bool {
        use TokenType::*;

        if let Some(t) = self.tokens.last() {
            matches!(
                t.tt,
                Bool(_)
                    | Char(_)
                    | CloseBrace
                    | CloseParen
                    | CloseBracket
                    | Ident(_)
                    | Num(_)
                    | Op(Operator::Inc)
                    | Op(Operator::Dec)
            )
        } else {
            false
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, Clone, Copy, PartialEq)]
struct ContextElement<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextElement<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextElement { value, line: line + 1, column: column + 1 }
    }
}

impl ContextElement<char> {
    // Will cause lexing to stop if there's a null byte in the file
    fn is_eof(&self) -> bool {
        self.value == 0 as char
    }
}

impl PartialEq<char> for ContextElement<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextElement<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let opt = self.lines.get(self.line);
        let line = match opt {
            Some(l) => l,
            None => return Some(ContextElement::new(0 as char, self.line, self.column - 1)),
        };
        let cc =
            line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column)).or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column).map(|c| ContextElement::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc.or_else(|| Some(ContextElement::new(0 as char, self.line, self.column - 1)))
    }
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Lexing error: {} at {}:{}", self.message, self.line, self.column)
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextElement<T>)> for LexError {
    fn from((msg, cp): (String, ContextElement<T>)) -> Self {
        LexError { message: msg, line: cp.line, column: cp.column }
    }
}
