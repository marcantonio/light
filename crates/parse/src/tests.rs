use super::*;
use lex::Lex;

fn ast_to_string(ast: Result<&Ast<ast::Node>, &ParseError>) -> String {
    match ast {
        Ok(ast) => ast.nodes().iter().map(|x| x.to_string()).collect(),
        Err(err) => err.to_string(),
    }
}

macro_rules! run_insta {
    ($prefix:expr, $tests:expr) => {
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for test in $tests {
                let tokens = Lex::new(test[1]).scan().expect("lexing failed in `parse` tests");
                let mut symbol_table = SymbolTable::new();
                let ast = Parse::new(&tokens, &mut symbol_table).parse().map(|(ast, ..)| ast);
                let ast_string = ast_to_string(ast.as_ref());
                insta::assert_yaml_snapshot!(format!("{}_{}", $prefix, test[0]), (test[1], ast, ast_string));
            }
        })
    };
}

#[test]
fn test_assignment() {
    let tests = [["basic", "x = x + 1"]];
    run_insta!("assignment", tests);
}

#[test]
fn test_binop() {
    let tests = [
        ["basic", "20 - 19"],
        ["invalid_op", "20 ! 19"],
        ["basic_var", "19 + a + 40"],
        ["bad_1", "20 -"],
        ["bad_2", "20 - else"],
        ["empty_paren", "()"],
        ["bad_3", "("],
        ["bad_4", "(19 + 21"],
        ["multi", "19 ** 21 ** 40; 19 - 21 * 20 + 40"],
        ["with_unop", "19 + 21 + -40"],
        ["single", "19"],
    ];
    run_insta!("binop", tests);
}

#[test]
fn test_block() {
    let tests = [
        ["basic", "if foo { bar }"],
        ["empty", "if foo {}"],
        ["missing", "if foo"],
        ["bad", "if foo {"],
        ["let", "let x: int = { 1 }"],
        ["multi", "y = { 1 + 4; foo(); 4 / 2 }"],
    ];
    run_insta!("block", tests);
}

#[test]
fn test_call() {
    let tests = [
        ["basic", "a(b, c)"],
        ["bad_1", "a("],
        ["bad_2", "a(a"],
        ["bad_3", "a(a,"],
        ["bad_4", "a(b b)"],
        ["nested", "a(b(c))"],
        ["nullary", "a()"],
        ["with_expr", "a(b + 1, c - 2 / 4)"],
    ];
    run_insta!("call", tests);
}

#[test]
fn test_char() {
    let tests = [["basic", "let c: char = 'a'"], ["control", "let nl: char = '\n'"]];
    run_insta!("char", tests);
}

#[test]
fn test_extern() {
    let tests = [["basic", "extern fn cos(x: float)"], ["err", "extern cos(x: float)"]];
    run_insta!("extern", tests);
}

#[test]
fn test_for() {
    let tests = [
        ["basic", "for x: int = 1; x < 10; 1 { foo }"],
        ["bad_1", "for"],
        ["bad_2", "for a"],
        ["bad_3", "for a:"],
        ["bad_4", "for a: int"],
        ["bad_5", "for a: int ="],
        ["bad_6", "for a: int = 1"],
    ];
    run_insta!("for", tests);
}

#[test]
fn test_func_def() {
    let tests = [
        [
            "basic",
            r#"
fn a(b: int, c: int) -> int {
    19 + a + 40
}
"#,
        ],
        [
            "multi_expr",
            r#"
fn a(b: int, c: int) -> int {
    19 + a + 40
    b + a
}
"#,
        ],
        [
            "no_ret_ty",
            r#"
fn a(b: int) {
    b
}
"#,
        ],
        ["empty", "fn a() {}"],
        [
            "multi_arg",
            r#"
fn (a: int, c: int) {
    1 + 2
}
"#,
        ],
        [
            "redefine_fn",
            r#"
fn foo() {}
fn bar() {}
fn foo() {}
"#,
        ],
        ["bad_1", "fn a(b: int) - { b }"],
        ["bad_2", "fn a(b: int) -> { b }"],
        ["bad_3", "fn a(b: int) -> not_a_type { b }"],
        ["bad_4", "fn a"],
        ["bad_5", "fn a( {}"],
        ["bad_6", "fn a(a: int, {}"],
        ["bad_7", "fn a(b: int b: int)"],
        ["bad_8", "fn a("],
        ["bad_9", "fn a(a: int, {}"],
        ["bad_10", "fn a(a: int"],
        ["bad_11", "fn a(a: int,"],
        ["bad_12", "fn a(b:)"],
        ["bad_13", "fn a(b)"],
    ];
    run_insta!("func_def", tests);
}

#[test]
fn test_gt_lt() {
    let tests = [["gt", "a + b > 40"], ["lt", "a < b / 40"]];
    run_insta!("gt_lt", tests);
}

#[test]
fn test_if() {
    let tests = [
        ["then", "if a > b { foo }"],
        [
            "else",
            r#"
if a > b {
    foo
} else {
    bar
}"#,
        ],
        [
            "then_else_if",
            r#"
if a > b {
    foo
} else if c < a {
    bar
}
"#,
        ],
        [
            "then_else_if_else",
            r#"
if a > b {
    foo
} else if c < a {
    bar
} else {
    baz
}
"#,
        ],
        ["bool_1", "if true { foo } else { bar }"],
        ["bool_2", "if false { foo } else { bar }"],
        ["bad", "if a > b { foo } else bar"],
    ];
    run_insta!("if", tests);
}

#[test]
fn test_let() {
    let tests = [
        ["basic", "let x: int = 1"],
        ["basic_float", "let x: float = 1.0"],
        ["bad_1", "let x: int ="],
        ["bad_2", "let x: int"],
        ["bad_3", "let x:"],
        ["bad_4", "let x"],
        ["bad_5", "let"],
    ];
    run_insta!("let", tests);
}

#[test]
fn test_logical_ops() {
    let tests = [
        ["eq", "x == 1"],
        ["and", "x && 1"],
        ["or", "x || 1"],
        ["complex", "x == 1 && y == 2 || z == 3"],
        ["complex_parens", "x == 1 && (y == 2 || z == 3)"],
    ];
    run_insta!("logical_ops", tests);
}

#[test]
fn test_op_prec() {
    let tests = [
        ["left_1", "19 + 21 * 40"],
        ["left_2", "19 * 21 - 40"],
        ["left_3", "19 - 21 + 40"],
        ["left_4", "19 - 21 * 20 + 40"],
        ["right_1", "19 ** 21 ** 40"],
        ["right_2", "19 ** 21 + 40"],
        ["right_3", "19 ** 21 ** 40 / 2"],
    ];
    run_insta!("op_prec", tests);
}

#[test]
fn test_parens() {
    let tests =
        [["excessive", "(((0)))"], ["prec", "(19 + 21) / 40"], ["complex", "3 * ((19 + 21) - 5) / 40"]];
    run_insta!("parens", tests);
}

#[test]
fn test_unop() {
    let tests = [
        ["basic", "-21"],
        ["with_binop", "-a * 2"],
        ["sub", "3 - -21"],
        ["right", "-4 ** 2"],
        ["double_neg_good", "-(-21)"],
        ["double_neg_bad", "--21"],
        ["invalid", "*2"],
    ];
    run_insta!("unop", tests);
}

#[test]
fn test_array() {
    let tests = [
        ["type", "let x: [int; 3]"],
        ["type_bad_1", "let x: [int;"],
        ["type_bad_2", "let x: [int"],
        ["type_bad_3", "let x: [int]"],
        ["type_bad_4", "let x: ["],
        ["type_bad_5", "let x: [foo]"],
        ["lit", "let x: [int; 3] = [1, 2, 3]"],
        ["empty_lit", "let x: [int; 3] = []"],
        ["index_1", "x[0]"],
        ["index_2", "x[1 + 2]"],
        ["index_3", "x[y[0]]"],
        ["index_bad_1", "x["],
        ["index_bad_2", "x[]"],
        ["index_bad_3", "x[0"],
        ["index_chain", "x[0][1]"],
    ];
    run_insta!("array", tests)
}

#[test]
fn test_struct() {
    let tests = [
        [
            "fields_only",
            r#"
struct Foo {
    let a: int
    let b: float = 1.0
}
"#,
        ],
        [
            "methods_only",
            r#"
struct Foo {
    fn c() -> int {
        1
    }
    fn d() {}
}
"#,
        ],
        [
            "mix",
            r#"
struct Foo {
    let a: int
    let b: float = 1.0

    fn c() -> int {
        1
    }
    fn d() {}
}
"#,
        ],
        [
            "dup",
            r#"
struct Foo {}
struct Foo {}
"#,
        ],
        ["field_selector", "x.y"],
        ["field_selector_bad", "x."],
        ["method_selector", "x.y(); x.y(a); x.y(a, b)"],
        ["method_selector_bad1", "x.y("],
        ["field_selector_bad2", "x.y(a"],
        ["field_selector_bad3", "x.y(a,"],
        ["field_selector_bad4", "x.y(a a"],
        [
            "selector_chaining",
            r#"
x.x.x
x().y()
x.y.z()
x.y().z
"#,
        ],
        ["empty", "struct Foo {}"],
        ["single_line_struct_with_field", "struct Foo { let x: int }"],
        ["single_line_struct_with_method", "struct Foo { fn foo() {} }"],
    ];
    run_insta!("struct", tests)
}
