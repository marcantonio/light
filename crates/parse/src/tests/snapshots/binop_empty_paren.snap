---
source: crates/parse/src/tests.rs
expression: "(test[1], ast, ast_string)"
---
- ()
- Err:
    - message: "Expecting primary expression. Got `CloseParen`"
      line: 1
      column: 2
- "Expecting primary expression. Got `CloseParen` at 1:2"

